src/cfg.rs

```diff
@@ -415,124 +400,123 @@ impl SExpr {
     }
 }
 
-/// Get the root expressions and strip comments.
-fn get_root_exprs(cfg: &str) -> Result<Vec<String>> {
-    let mut open_paren_count = 0;
-    let mut close_paren_count = 0;
-    let mut s_exprs = Vec::new();
-    let mut cur_expr = String::new();
-    for line in cfg.lines() {
-        // remove comments
-        let line = line.split(";;").next().unwrap();
-        for c in line.chars() {
-            if c == '(' {
-                open_paren_count += 1;
-            } else if c == ')' {
-                close_paren_count += 1;
+#[derive(Debug)]
+enum Token<'a> {
+    Open,
+    Close,
+    String(&'a str),
+}
+
+struct Lexer<'a> {
+    s: &'a str,
+    bytes: Bytes<'a>,
+}
+
+fn is_start(b: u8) -> bool {
+    matches!(b, b'(' | b')' | b'"') || b.is_ascii_whitespace()
+}
+
+impl<'a> Lexer<'a> {
+    fn new(s: &str) -> impl Iterator<Item = Result<Token<'_>>> {
+        let mut lexer = Lexer {
+            s,
+            bytes: s.bytes(),
+        };
+        iter::from_fn(move || lexer.next_token())
+    }
+
+    fn next_while(&mut self, f: impl Fn(u8) -> bool) {
+        while let Some(b) = self.bytes.clone().next() {
+            if f(b) {
+                self.bytes.next().unwrap();
+            } else {
+                break;
             }
         }
-        if open_paren_count == 0 {
-            continue;
-        }
-        cur_expr.push_str(line);
-        cur_expr.push('\n');
-        if open_paren_count == close_paren_count {
-            open_paren_count = 0;
-            close_paren_count = 0;
-            s_exprs.push(cur_expr.trim().to_owned());
-            cur_expr.clear();
-        }
     }
-    if !cur_expr.is_empty() {
-        bail!("Unclosed root expression:\n{}", cur_expr)
+
+    fn pos(&self) -> usize {
+        self.s.len() - self.bytes.len()
+    }
+
+    fn consumed(&self, start: usize) -> &'a str {
+        &self.s[start..self.pos()]
     }
-    Ok(s_exprs)
-}
 
-/// Parse an expression string into an SExpr
-fn parse_expr(expr: &str) -> Result<Vec<SExpr>> {
-    if !expr.starts_with('(') {
-        bail!("Expression in cfg does not start with '(':\n{}", expr)
+    fn next_token(&mut self) -> Option<Result<Token<'a>>> {
+        use Token::*;
+        loop {
+            let start = self.pos();
+            break match self.bytes.next() {
+                Some(b) => Some(Ok(match b {
+                    b'(' => Open,
+                    b')' => Close,
+                    b'"' => {
+                        self.next_while(|b| b != b'"' && b != b'\n');
+                        match self.bytes.next() {
+                            Some(b'"') => (),
+                            _ => return Some(Err(anyhow!("Unterminated string"))),
```

Comment:
A return type of `Result<Option<Token<'a>>>` would allow this line to just be `_ => bail!(...)`. It doesn't seem like the rest of the function would be too badly affected by this change, so should consider changing the return type.

Reply:
I did this so that it fits with an iterator easily. It's not much more code than using `bail` anyway